src/setup_payload/Base38.cpp

```diff
@@ -17,35 +17,34 @@
 
 /**
  *    @file
- *      This file implements converting an array of bytes into a Base41 String and
+ *      This file implements converting an array of bytes into a Base38 String and
  *       the reverse.
  *
  *      The encoding chosen is: treat every 2 bytes of input data as a little-endian
- *        uint16_t, then div and mod that into 3 base41 characters, with the least-significant
+ *        uint16_t, then div and mod that into 3 base38 characters, with the least-significant
  *        encoding bits in the first character of the resulting string.  If an odd
  *        number of bytes is encoded, 2 characters are used to encode the last byte.
  *
  */
 
-#include "Base41.h"
+#include "Base38.h"
 
 #include <climits>
 
 namespace chip {
 
-static const char codes[]        = { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D',
-                              'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R',
-                              'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', ' ', '+', '-', '.', '/' };
-static const int kBase41ChunkLen = 3;
+static const char codes[]        = { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I',
+                              'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', '-', '.' };
+static const int kBase38ChunkLen = 3;
 static const int kBytesChunkLen  = 2;
 static const int kRadix          = sizeof(codes) / sizeof(codes[0]);
 
-std::string base41Encode(const uint8_t * buf, size_t buf_len)
+std::string base38Encode(const uint8_t * buf, size_t buf_len)
 {
     std::string result;
 
     // eat little-endian uint16_ts from the byte array
-    // encode as 3 base41 characters
```

Comment:
I think this is wrong now as `38^3 == 54872 < 2^16`. The spec suggests encoding every 3 bytes using 5 characters. I'm surprised we don't have unit tests which caught this. We should probably add some tests which do encoding of payload with FF bytes, decoding, and verifying that the contents didn't change.

Reply:
If the spec is strictly followed then groups of 1 and 3 characters will be undecodable since it only mentions groups of 2, 4 and 5. The previous code had a size optimization which allowed shortening of codes (which is not a bad idea with QR codes) and could well be applied here as well, meaning there will be no undecodable sequence of allowed characters. What that means for base38: encoding 1 byte needs 1 or 2 characters, encoding 2 bytes needs 2, 3, or 4 characters and encoding 3 bytes needs 4 or 5 characters.

38                  0x00000026
38*38               0x000005a4
38*38*38            0x0000d658
38*38*38*38         0x001fd110
38*38*38*38*38      0x04b90860

Reply:
You're. I don't think we need to care about input strings which couldn't be generated by a CHIP-compliant encoder, but it's a valid point that the optimization might have been proposed. Looking at the change: https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/2267, it appears that the optimization hadn't been part of the spec even before, am I right?

@jrhees-cae, has the size optimization for last bytes of a QR Code payload ever been considered? Do you think it's worth adding to the spec?

Reply:
@Damian-Nordic The previous optimization (that was coded in the Base-41 implementation) was not in the spec. I believe is was added by the original implementer. We *could* add such an optimization to the spec, but there was no clear consensus as to whether the complexity (added code and description) was warranted for the size savings.

Reply:
@jrhees-cae Thanks for your answer. I believe that for this PR the optimization should be removed from the code until it gets discussed and accepted in the spec then.

@bluebin14 Is the size optimization very important for you? If so, do you think you could describe your use case in the spec repo and propose the solution?

Reply:
There is actually a 100% use case for the optimization, without it all CHIP QR codes will always have useless 2 zeros at the end: CH:-AM148DB00CO3Z5X900 instead of CH:-AM148DB00CO3Z5X9. The length in bytes is 88 bits which is 11 bytes which is 3 x 3 + 2 so you always end up with the 2 byte case and encode it with 4 characters, the last 2 being 0.

Reply:
@bluebin14 While this is certainly true, in practice this does not cause a change in the rendered QR code size (since it does not cause a shift to a higher level code)

Reply:
Added issue #2413 to spec repo with solution.

Reply:
@bluebin14 I guess you meant wrong issue in the comment and it should be that one: https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/2413 nevertheless thank you for creating it. I guess for now consensus it to implement code without optimization and align it after potential spec update, right?

Reply:
That is exactly what I meant by "issue #2413 to spec repo". I wasn't sure if it was ok to post links to private repos.

Reply:
Added discussion label to this. It can be discussed at next QR Code spec meeting